### DISCUSSION TITLE
drivebase reset function

### DISCUSSION THREAD

**Discussion by nothingtodowu** [pybricks/support#1616] [General]
When I write my program at Spike primehub with pybricks, It is amazing that I can go straight 2000 mm and it is no any wrong when I enable use_gyro(True).  But I have an issue that when I have more and more operations like robot.straight/robot.turn, then suddenly something went wrong. I turn 90 degree but it turn randomly, maybe 85 degree maybe 95 degree. I try to add robot.reset function to reset all distance and angle, but it still same. After I power off/on my primehub, it works better. But after more operations, it comes again. 

Is there any function in Pybricks to reset the drivebase to initial? Thanks a lot.

**Comment by laurensvalk** [comment]
Which stop mode are you using?

You should use Hold for best performance (this is the default). This keeps error accumulation minimal.

> randomly

Is it random, or is it a constant offset/error?

The gyros are not a compass, so they tend to have a slight offset. You can test this with your hub by putting it flat on the table against a wall. Then rotate it. Maybe it will always say 362, for example. Once you know this offset, you can take it into account in your code.

> more and more operations

The gyro is pretty good, but it's still recommend to also use other sensors to realign yourself.

**Reply by nothingtodowu** [reply]
Thanks a lot. At first everything working fine, even go srtaight 2000 mm still no issue. But after multiple straight/turn then it goes wrong. It can't exactly turn 90 degree. I think it is randomly because sometime it turn 85 degree and sometime it turn 95 degree. I think it maybe error accumulation too much. I use color sensor and wall for calibrate my location, but robot can't turn accurate will be big issue. Is there anyway to re-initialise drivebase?

**Reply by laurensvalk** [reply]
> sometime it turn 85 degree and sometime it turn 95 degree.

So it ends up at the correct value again? This seems like the correct behavior.

> I use color sensor and wall for calibrate my location

After doing that, make sure to call the drivebase stop with coast or brake, so it won't try to maintain its original orientation.

**Reply by nothingtodowu** [reply]
I just use default stop mode. How to avoid this "error accumulation"? Any reset/re-initialize function? Now I just use power loop to fix this issue. But it really waste time.

**Comment by ggramlich** [comment]
@nothingtodowu Are you aware of https://github.com/orgs/pybricks/discussions/1595#discussioncomment-9103246 ?

The issues that you are describing sound similar to ours. They were resulting from us thinking that we could use the options "then hold", "then coast", "then brake" almost interchangeably, but they make a huge difference, especially if you mix them.

**Reply by nothingtodowu** [reply]
Thanks a lot! I am also not sure how to use hold/coast/coast_smart/brake. I just use default Stop.Hold mode. In motor there is run_target(speed, target_angle, then=Stop.HOLD, wait=True) function to let motor run better than run_angle(speed, rotation_angle, then=Stop.HOLD, wait=True). This way really can avoid error accumulation. But for drivebase, there is only robot.turn function. How did you fix this issue when you have similar one? Actually for me temporary solution is when the issue happen, just power loop the hub. Most of time can solve it. But it waste time.

**Reply by ggramlich** [reply]
> How did you fix this issue when you have similar one?

As I wrote https://github.com/orgs/pybricks/discussions/1595 we more consistently used hold and took special care of the parts where we align against a wall by switching off and on the useGyro. Also see my comment on https://github.com/pybricks/support/issues/1617.
As @MrGibbage wrote: It would be best to see a video of your robot's behavior.

**Comment by MrGibbage** [comment]
I'm just an end user here, but I would like to see a video of what you are experiencing. We used pybricks on Spike Prime last year, and every time we thought we were having a pybricks problem, it turned out to always be a problem that we had in our code, or a mechanical problem with the robot itself. But we are always on the lookout for the true limitations. So far we felt like we were pushing the envelope with long runs, but the robot has never really let us down. Seeing a video of your robot misbehaving could be really helpful not just to the developers here, but to other end users as well.

**Comment by rajeshsubramanianus** [comment]
@laurensvalk Happy New Year! I got similar issue as mentioned by users above. I use Spike prime hub and Pybricks block coding. The code after the comment in the below screenshot doesn't work. When I checked the robot, the wheels got stuck and when I lifted the robot from the field they are rotating in the opposite direction. We have to participate in the state level competition soon, so any inputs are greatly appreciated. I used "Continue" "Hold" "Brake" in the command before comment, but no luck.

- **UPDATE**: The above issue is resolved. Found out the last command after which robot got stuck couldn't fully execute. That command was a anti-clockwise motor run for 220 degrees, but there was a Lego piece that was blocking it on the end (may be at 200 degree) and not allowing it to complete the full 220 degrees run. After removing that lego piece from the path, it is now working perfectly!

![image](https://github.com/user-attachments/assets/b8ce1008-aa0c-4cb9-918e-6eae324d7bfc)

**Reply by BertLindeman** [reply]
Thank you for the update! 

I could not reconstruct your problem.
Glad you fixed it.

Bert

**Reply by rajeshsubramanianus** [reply]
Thanks for the quick reply. 👍
